File: app/channels/telegram.py
"""Telegram implementation of :class:`MessagingChannel`.

Uses ``python-telegram-bot`` in webhook mode. Inbound updates arrive at the
FastAPI webhook; we verify the secret-token header there, then hand the raw
payload to :meth:`parse_update`.
"""
from __future__ import annotations

import io

from telegram import Bot, InlineKeyboardButton, InlineKeyboardMarkup, Update
from telegram.error import TimedOut
from telegram.request import HTTPXRequest

from app.channels.base import InboundMessage, InboundPhoto, MessagingChannel
from app.config import settings
from app.logging_config import get_logger
from app.text_style import strip_emoji

log = get_logger(__name__)


class TelegramChannel(MessagingChannel):
    name = "telegram"

    def __init__(self, token: str | None = None) -> None:
        self._token = token or settings.telegram_bot_token
        self._bot: Bot | None = None
        self._initialized = False

    @property
    def enabled(self) -> bool:
        return bool(self._token)

    @property
    def bot(self) -> Bot:
        if not self._token:
            raise RuntimeError("TELEGRAM_BOT_TOKEN is not configured")
        if self._bot is None:
            # Generous pool for concurrent sends. Telegram media fetches can be
            # noticeably slower than text requests, so use roomier timeouts.
            request = HTTPXRequest(
                connection_pool_size=8,
                connect_timeout=10.0,
                read_timeout=30.0,
                write_timeout=30.0,
                pool_timeout=10.0,
            )
            self._bot = Bot(token=self._token, request=request)
        return self._bot

    async def initialize(self) -> None:
        if self.enabled and not self._initialized:
            await self.bot.initialize()
            self._initialized = True

    async def _ensure_ready(self) -> None:
        """Lazily initialize on first use so a startup-time connectivity blip
        doesn't permanently disable the transport."""
        if not self._initialized:
            await self.initialize()

    async def shutdown(self) -> None:
        if self._bot is not None:
            await self._bot.shutdown()
            self._initialized = False

    # ----- outbound -----
    # Emoji is stripped here as the final transport-level guard, so no outbound
    # text or caption can carry emoji regardless of its source (LLM, canned,
    # captions passed by callers).
    async def send_text(self, chat_id: int, text: str) -> None:
        await self._ensure_ready()
        text = strip_emoji(text)
        # Telegram hard-limits messages to 4096 chars; chunk long replies.
        for chunk in _chunk(text, 4096):
            await self.bot.send_message(chat_id=chat_id, text=chunk)

    async def send_text_with_keyboard(
        self, chat_id: int, text: str, buttons: list[list[tuple[str, str]]]
    ) -> None:
        await self._ensure_ready()
        text = strip_emoji(text)
        markup = InlineKeyboardMarkup(
            [
                [InlineKeyboardButton(label, callback_data=data) for label, data in row]
                for row in buttons
            ]
        )
        await self.bot.send_message(chat_id=chat_id, text=text, reply_markup=markup)

    async def answer_callback(self, callback_query_id: str) -> None:
        await self._ensure_ready()
        try:
            await self.bot.answer_callback_query(callback_query_id)
        except Exception:  # pragma: no cover - non-fatal spinner cleanup
            log.debug("answer_callback_query failed", exc_info=True)

    async def send_photo(
        self, chat_id: int, file_id_or_bytes, caption: str | None = None
    ) -> None:
        await self._ensure_ready()
        await self.bot.send_photo(
            chat_id=chat_id,
            photo=file_id_or_bytes,
            caption=strip_emoji(caption) if caption else caption,
        )

    async def send_document(
        self, chat_id: int, data: bytes, filename: str, caption: str | None = None
    ) -> None:
        await self._ensure_ready()
        bio = io.BytesIO(data)
        bio.name = filename
        await self.bot.send_document(
            chat_id=chat_id,
            document=bio,
            filename=filename,
            caption=strip_emoji(caption) if caption else caption,
        )

    async def download_photo(self, file_id: str) -> bytes:
        await self._ensure_ready()
        for attempt in range(2):
            try:
                tg_file = await self.bot.get_file(file_id)
                buf = bytearray()
                await tg_file.download_as_bytearray(buf)
                return bytes(buf)
            except TimedOut:
                if attempt == 1:
                    raise
                log.warning("Telegram photo download timed out; retrying once")
        raise RuntimeError("unreachable")

    async def set_webhook(self) -> bool:
        await self._ensure_ready()
        return await self.bot.set_webhook(
            url=settings.webhook_url,
            secret_token=settings.telegram_webhook_secret,
            allowed_updates=["message", "callback_query"],
            drop_pending_updates=True,
        )

    async def delete_webhook(self) -> bool:
        await self._ensure_ready()
        return await self.bot.delete_webhook(drop_pending_updates=False)

    # ----- inbound -----
    def parse_update(self, update: dict) -> InboundMessage | None:
        try:
            upd = Update.de_json(update, self.bot if self._token else None)
        except Exception:  # pragma: no cover - defensive
            log.exception("Failed to parse Telegram update")
            return None

        # Inline-button tap (portion confirmation).
        if upd.callback_query is not None and upd.callback_query.from_user is not None:
            cq = upd.callback_query
            chat_id = cq.message.chat_id if cq.message is not None else cq.from_user.id
            return InboundMessage(
                channel=self.name,
                external_user_id=cq.from_user.id,
                chat_id=chat_id,
                text=cq.data,
                callback_data=cq.data,
                callback_query_id=cq.id,
                username=cq.from_user.username,
                first_name=cq.from_user.first_name,
                raw=update,
            )

        msg = upd.message or upd.edited_message
        if msg is None or msg.from_user is None:
            return None

        photo: InboundPhoto | None = None
        if msg.photo:
            # Telegram sends multiple sizes; the last is the largest.
            largest = msg.photo[-1]
            photo = InboundPhoto(
                file_id=largest.file_id,
                file_unique_id=largest.file_unique_id,
                width=largest.width,
                height=largest.height,
            )

        return InboundMessage(
            channel=self.name,
            external_user_id=msg.from_user.id,
            chat_id=msg.chat_id,
            text=msg.text,
            photo=photo,
            caption=msg.caption,
            username=msg.from_user.username,
            first_name=msg.from_user.first_name,
            raw=update,
        )


def _chunk(text: str, size: int):
    if not text:
        yield ""
        return
    for i in range(0, len(text), size):
        yield text[i : i + size]

File: src/channels/telegram.ts
import type { Settings } from "../config";
import { stripEmoji } from "../services/text-style";
import type {
  ButtonRow,
  DownloadedImage,
  InboundMessage,
  InboundPhoto,
  MessagingChannel,
  SendDocumentOptions,
  SendPhotoOptions,
} from "./types";

const TELEGRAM_CAPTION_MAX = 1024;

const MIME_BY_EXT: Record<string, string> = {
  jpg: "image/jpeg",
  jpeg: "image/jpeg",
  png: "image/png",
  webp: "image/webp",
  gif: "image/gif",
  heic: "image/heic",
};

function detectMimeFromPath(path: string, contentType: string | null): string {
  if (contentType && contentType.startsWith("image/")) {
    return contentType.split(";")[0]!.trim();
  }
  const ext = path.split(".").pop()?.toLowerCase() ?? "";
  return MIME_BY_EXT[ext] ?? "image/jpeg";
}

function chunkText(text: string, size: number): string[] {
  if (!text) return [""];
  const chunks: string[] = [];
  for (let i = 0; i < text.length; i += size) {
    chunks.push(text.slice(i, i + size));
  }
  return chunks;
}

export class TelegramChannel implements MessagingChannel {
  readonly name = "telegram";

  constructor(
    private readonly settings: Settings,
    private readonly token?: string,
  ) {}

  get enabled(): boolean {
    return Boolean(this.token ?? this.settings.telegramBotToken);
  }

  private get botToken(): string {
    const token = this.token ?? this.settings.telegramBotToken;
    if (!token) throw new Error("TELEGRAM_BOT_TOKEN is not configured");
    return token;
  }

  private get baseUrl(): string {
    return `https://api.telegram.org/bot${this.botToken}`;
  }

  private get fileUrl(): string {
    return `https://api.telegram.org/file/bot${this.botToken}`;
  }

  private isBenignTelegramError(description: string): boolean {
    return (
      description.includes("message is not modified") ||
      description.includes("query is too old") ||
      description.includes("query ID is invalid")
    );
  }

  private async call(method: string, payload: Record<string, unknown>): Promise<unknown> {
    const resp = await fetch(`${this.baseUrl}/${method}`, {
      method: "POST",
      headers: { "content-type": "application/json" },
      body: JSON.stringify(payload),
    });
    const data = (await resp.json()) as { ok?: boolean; result?: unknown; description?: string };
    if (!data.ok) {
      const detail = data.description ?? `Telegram ${method} failed`;
      if (this.isBenignTelegramError(detail)) {
        return data.result ?? null;
      }
      throw new Error(detail);
    }
    return data.result;
  }

  async sendText(
    chatId: string | number,
    text: string,
    _replyToken?: string | null,
    parseMode?: "HTML",
  ): Promise<void> {
    await this.sendTextReturningId(chatId, text, parseMode);
  }

  async sendTextReturningId(
    chatId: string | number,
    text: string,
    parseMode?: "HTML",
  ): Promise<number | null> {
    const cleaned = stripEmoji(text);
    const chunks = chunkText(cleaned, 4096);
    let lastMessageId: number | null = null;
    for (const chunk of chunks) {
      const payload: Record<string, unknown> = { chat_id: chatId, text: chunk };
      if (parseMode) {
        payload.parse_mode = parseMode;
      }
      const result = (await this.call("sendMessage", payload)) as {
        message_id?: number;
      };
      lastMessageId = result.message_id ?? lastMessageId;
    }
    return lastMessageId;
  }

  async deleteMessage(
    chatId: string | number,
    messageId: string | number,
  ): Promise<void> {
    try {
      await this.call("deleteMessage", {
        chat_id: chatId,
        message_id: Number(messageId),
      });
    } catch {
      // Non-fatal if already deleted or too old.
    }
  }

  async sendTextWithKeyboard(
    chatId: string | number,
    text: string,
    buttons: ButtonRow[],
    _replyToken?: string | null,
    parseMode?: "HTML",
  ): Promise<number | null> {
    const replyMarkup = {
      inline_keyboard: buttons.map((row) =>
        row.map((btn) => ({ text: btn.label, callback_data: btn.data })),
      ),
    };
    const payload: Record<string, unknown> = {
      chat_id: chatId,
      text: stripEmoji(text),
      reply_markup: replyMarkup,
    };
    if (parseMode) {
      payload.parse_mode = parseMode;
    }
    const result = (await this.call("sendMessage", payload)) as {
      message_id?: number;
    };
    return result.message_id ?? null;
  }

  async editMessageReplyMarkup(
    chatId: string | number,
    messageId: string | number,
    buttons: ButtonRow[],
  ): Promise<void> {
    const replyMarkup = {
      inline_keyboard: buttons.map((row) =>
        row.map((btn) => ({ text: btn.label, callback_data: btn.data })),
      ),
    };
    await this.call("editMessageReplyMarkup", {
      chat_id: chatId,
      message_id: Number(messageId),
      reply_markup: replyMarkup,
    });
  }

  async editMessageText(
    chatId: string | number,
    messageId: string | number,
    text: string,
    parseMode?: "HTML",
    buttons?: ButtonRow[],
  ): Promise<void> {
    const payload: Record<string, unknown> = {
      chat_id: chatId,
      message_id: Number(messageId),
      text: stripEmoji(text),
    };
    if (parseMode) {
      payload.parse_mode = parseMode;
    }
    if (buttons) {
      payload.reply_markup = {
        inline_keyboard: buttons.map((row) =>
          row.map((btn) => ({ text: btn.label, callback_data: btn.data })),
        ),
      };
    }
    await this.call("editMessageText", payload);
  }

  async clearMessageReplyMarkup(
    chatId: string | number,
    messageId: string | number,
  ): Promise<void> {
    await this.call("editMessageReplyMarkup", {
      chat_id: chatId,
      message_id: Number(messageId),
      reply_markup: { inline_keyboard: [] },
    });
  }

  async answerCallback(
    callbackQueryId: string,
    options?: { text?: string },
  ): Promise<void> {
    try {
      const payload: Record<string, unknown> = {
        callback_query_id: callbackQueryId,
      };
      if (options?.text) {
        payload.text = options.text;
        payload.show_alert = false;
      }
      await this.call("answerCallbackQuery", payload);
    } catch {
      // Non-fatal: spinner may already be cleared
    }
  }

  async sendPhoto(
    chatId: string | number,
    options: SendPhotoOptions,
  ): Promise<void> {
    const photo = options.fileId || options.imageUrl;
    if (!photo) {
      throw new Error("Telegram sendPhoto requires fileId or imageUrl");
    }

    const payload: Record<string, unknown> = {
      chat_id: chatId,
      photo,
    };
    if (options.caption) {
      const caption = stripEmoji(options.caption).slice(0, TELEGRAM_CAPTION_MAX);
      payload.caption = caption;
      if (options.parseMode) {
        payload.parse_mode = options.parseMode;
      }
    }
    await this.call("sendPhoto", payload);
  }

  async sendDocument(
    chatId: string | number,
    options: SendDocumentOptions,
  ): Promise<void> {
    if (options.bytes && options.bytes.byteLength > 0) {
      const form = new FormData();
      form.append("chat_id", String(chatId));
      form.append(
        "document",
        new Blob([options.bytes], {
          type: options.mimeType || "application/pdf",
        }),
        options.filename,
      );
      if (options.caption) {
        form.append("caption", stripEmoji(options.caption));
      }
      const resp = await fetch(`${this.baseUrl}/sendDocument`, {
        method: "POST",
        body: form,
      });
      const data = (await resp.json()) as {
        ok?: boolean;
        description?: string;
      };
      if (!data.ok) {
        throw new Error(data.description ?? "Telegram sendDocument failed");
      }
      return;
    }

    if (!options.fileUrl) {
      throw new Error("Telegram sendDocument requires bytes or fileUrl");
    }
    const payload: Record<string, unknown> = {
      chat_id: chatId,
      document: options.fileUrl,
    };
    if (options.caption) {
      payload.caption = stripEmoji(options.caption);
    }
    await this.call("sendDocument", payload);
  }

  async downloadPhoto(fileId: string): Promise<DownloadedImage> {
    const result = (await this.call("getFile", { file_id: fileId })) as {
      file_path?: string;
    };
    const path = result.file_path;
    if (!path) {
      throw new Error("Telegram file_path missing");
    }

    const resp = await fetch(`${this.fileUrl}/${path}`);
    if (!resp.ok) {
      throw new Error(`Telegram file download failed: ${resp.status} path=${path}`);
    }

    const mime = detectMimeFromPath(path, resp.headers.get("content-type"));
    const bytes = new Uint8Array(await resp.arrayBuffer());
    if (bytes.byteLength === 0) {
      throw new Error(`Telegram file download empty: path=${path}`);
    }

    return { bytes, mime };
  }

  async setWebhook(): Promise<boolean> {
    await this.call("setWebhook", {
      url: this.settings.webhookUrl,
      secret_token: this.settings.telegramWebhookSecret,
      allowed_updates: ["message", "callback_query"],
      drop_pending_updates: true,
    });
    return true;
  }

  async deleteWebhook(): Promise<boolean> {
    await this.call("deleteWebhook", { drop_pending_updates: false });
    return true;
  }

  parseUpdate(update: unknown): InboundMessage | null {
    if (!update || typeof update !== "object") return null;
    const u = update as Record<string, unknown>;

    const callback = u.callback_query;
    if (callback && typeof callback === "object") {
      const cb = callback as Record<string, unknown>;
      const fromUser = (cb.from ?? {}) as Record<string, unknown>;
      const message = (cb.message ?? {}) as Record<string, unknown>;
      const chat = (message.chat ?? {}) as Record<string, unknown>;
      return {
        channel: this.name,
        externalUserId: String(fromUser.id),
        chatId: Number(chat.id ?? fromUser.id),
        text: String(cb.data ?? ""),
        callbackData: String(cb.data ?? ""),
        callbackQueryId: String(cb.id ?? ""),
        callbackMessageId: message.message_id ? Number(message.message_id) : null,
        username: fromUser.username ? String(fromUser.username) : null,
        firstName: fromUser.first_name ? String(fromUser.first_name) : null,
        raw: update,
      };
    }

    const message = (u.message ?? u.edited_message) as Record<string, unknown> | undefined;
    if (!message || typeof message !== "object") return null;

    const fromUser = message.from as Record<string, unknown> | undefined;
    if (!fromUser?.id) return null;

    let photo: InboundPhoto | null = null;
    const photos = message.photo as Array<Record<string, unknown>> | undefined;
    if (photos?.length) {
      const largest = photos[photos.length - 1]!;
      photo = {
        fileId: String(largest.file_id),
        fileUniqueId: String(largest.file_unique_id),
        width: largest.width ? Number(largest.width) : undefined,
        height: largest.height ? Number(largest.height) : undefined,
      };
    }

    const document = message.document as Record<string, unknown> | undefined;
    if (!photo && document) {
      const mimeType = String(document.mime_type ?? "");
      if (mimeType.startsWith("image/")) {
        photo = {
          fileId: String(document.file_id),
          fileUniqueId: String(document.file_unique_id ?? document.file_id),
        };
      }
    }

    const chat = (message.chat ?? {}) as Record<string, unknown>;
    return {
      channel: this.name,
      externalUserId: String(fromUser.id),
      chatId: Number(chat.id ?? fromUser.id),
      text: message.text ? String(message.text) : null,
      photo,
      caption: message.caption ? String(message.caption) : null,
      username: fromUser.username ? String(fromUser.username) : null,
      firstName: fromUser.first_name ? String(fromUser.first_name) : null,
      raw: update,
    };
  }
}

export function createTelegramChannel(settings: Settings): TelegramChannel {
  return new TelegramChannel(settings);
}
